File: app/src/main/java/com/dev/fitface/ui/activity/CameraActivity.java
package com.dev.fitface.ui.activity;


import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;

import android.os.Bundle;

import android.os.CountDownTimer;
import android.util.Base64;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;


import com.dev.fitface.R;

public class CameraActivity extends AppCompatActivity {

    /*private static final int MY_PERMISSIONS_REQUEST_CAMERA = 1;
    private static final String TAG = "CAMERA ACTIVITY";

    RoundedFrameLayout roundedFrameLayout;
    TextView tvAction;
    CameraView cameraView;
    Fotoapparat fotoapparat;
    Boolean isRotatedRight, isRotatedLeft, isFaceProcessRunning;
    FirebaseVisionFaceDetector detector;
    long currentTime, prevTime;
    float minSmileProb, maxSmileProb;
    int leftF, topF, widthF, heightF;
    Bitmap bitmapF;
    String base64Str;
    Runnable runnable;
*/
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_camera);
  /*      setupUI();
        initServices();
        prevTime = System.currentTimeMillis();
        fotoapparat = startFotoapparat();
        checkPermission();
*/
    }
/*
    private void checkPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, MY_PERMISSIONS_REQUEST_CAMERA);
        } else {
            fotoapparat.start();
        }
    }

    private Fotoapparat startFotoapparat() {
        return Fotoapparat
                .with(this)
                .into(cameraView)
                .frameProcessor(frame -> {
                    currentTime = System.currentTimeMillis();
                    if (currentTime - 500 > prevTime) {
                        faceOptions(frame);
                        prevTime = currentTime;
                    }
                })
                .previewScaleType(ScaleType.CenterCrop)
                .lensPosition(front())
                .logger(loggers(
                        logcat(),
                        fileLogger(this)
                ))
                .cameraErrorCallback(e -> {
                    Toast.makeText(CameraActivity.this, e.toString(), Toast.LENGTH_LONG).show();
                })
                .photoResolution(standardRatio(
                        highestResolution()
                ))
                .build();

    }

    @Override
    protected void onPause() {
        super.onPause();
        fotoapparat.stop();
    }

    private void faceOptions(Frame frame) {
        detector.detectInImage(getVisionImageFromFrame(frame))
                .addOnSuccessListener(
                        faces -> {
 *//*                           // Kiểm tra có tồn tại duy nhất 1 khuôn mặt không
                            if (faces.size() > 1){
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Nhận diện lỗi");
                                });
                                tvAction.setText("Nhận diện lỗi");
                                return;
                            } else if(faces.size() < 1) {
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Nhận dạng thành công");
                                });
*//**//*                                tvAction.setText("Không tồn tại người nào ");
                                isFaceProcessRunning = true;*//**//*
                            }
                            FirebaseVisionFace face = faces.get(0);
                            Log.i(TAG, face.getBoundingBox().toString());
                            // Check front face
                            if (face.getHeadEulerAngleY() < -12 || face.getHeadEulerAngleY() > 12) {
                                isFaceProcessRunning = true;
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Vui lòng nhìn thẳng");
                                });
                            } else {
                                leftF = face.getBoundingBox().left;
                                topF = face.getBoundingBox().top;
                                widthF = face.getBoundingBox().width();
                                heightF = face.getBoundingBox().height();
//                                startCountdown(3);
                            }*//*


                            if (faces.size() > 1) {
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Có nhiều khuôn mặt");
                                });
                                isFaceProcessRunning = true;
                                return;
                            } else if (faces.size() < 1) {
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Không tìm thấy khuôn mặt");
                                });
                                isFaceProcessRunning = true;
                                return;
                            }
                            else if(faces.size() == 1){
                                FirebaseVisionFace face = faces.get(0);
                                Log.i(TAG, face.getBoundingBox().toString());
                                // Check front face
                                if (face.getHeadEulerAngleY() < -12 || face.getHeadEulerAngleY() > 12) {
                                    this.runOnUiThread(() -> {
                                        tvAction.setText("Vui lòng nhìn thẳng");
                                    });
                                    isFaceProcessRunning = true;
                                } else {
                                    this.runOnUiThread(() -> {
                                        tvAction.setText("Nhận diện thành công");
                                    });
                                    leftF = face.getBoundingBox().left;
                                    topF = face.getBoundingBox().top;
                                    widthF = face.getBoundingBox().width();
                                    heightF = face.getBoundingBox().height();
                                    isFaceProcessRunning = false;
                                    startCountdown(2);
                                }
                            }


                        *//*   // Check face sizeint
                          int left = roundedFrameLayout.getLeft();
                            int top = roundedFrameLayout.getTop();
                            int right = roundedFrameLayout.getRight();
                            int bottom = roundedFrameLayout.getBottom();

                            if (((leftF - left) > 0) && ((topF - top) > 0) && ((right - rightF) > 0) && (bottom - bottomF) > 0){
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Phát hiện khuôn mặt");
                                });
                            } else {
                                this.runOnUiThread(() -> {
                                    tvAction.setText("Không tìm thấy khuôn mặt");
                                });
                            }
                            return;*//*
                        }
                )
                .addOnFailureListener(e -> {
                });
    }


    private void startCountdown(int second) {
        *//*new CountDownTimer(second * 1000, 1000) {
            int time = second;
            public void onTick(long millisUntilFinished) {
                if (!isFaceProcessRunning) {
                    tvAction.setText("Bắt đầu chụp trong " + time + " giây nữa!");
                    time--;
                    if(time == 0){
                        takePhoto();
                    }
                }
            }

            public void onFinish() {
            }
        }.start();*//*
        new CountDownTimer(second * 1000, 1000){
            int time = second;

            @Override
            public void onTick(long millisUntilFinished) {
                tvAction.setText("Bắt đầu chụp trong " + time + " giây nữa!");
                time--;
            }

            @Override
            public void onFinish() {
                tvAction.setText("Bắt đầu chụp trong " + time + " giây nữa!");
                takePhoto();
            }
        }.start();
    }

    private void takePhoto() {
        final PhotoResult picture = fotoapparat.takePicture();
        fotoapparat.stop();
        fotoapparat.start();
        *//*picture.toBitmap().whenDone(bitmapPhoto -> {
            fotoapparat.stop();
            fotoapparat.start();
        });*//*
*//*        fotoapparat.stop();
        fotoapparat.start();*//*
    }

    private String bitmapToBase64(Bitmap bitmap){
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
        byte[] byteArray = byteArrayOutputStream .toByteArray();
        String encoded = Base64.encodeToString(byteArray, Base64.DEFAULT);
        return encoded;
    }

    private FirebaseVisionImage getVisionImageFromFrame(Frame frame) {
        byte[] bytes = frame.getImage();
        FirebaseVisionImageMetadata metadata = new FirebaseVisionImageMetadata.Builder()
                .setFormat(FirebaseVisionImageMetadata.IMAGE_FORMAT_NV21)
                .setRotation(FirebaseVisionImageMetadata.ROTATION_270)
                .setHeight(frame.getSize().height)
                .setWidth(frame.getSize().width)
                .build();

        FirebaseVisionImage image = FirebaseVisionImage.fromByteArray(bytes, metadata);
        return image;
    }

    private void initServices() {
        FirebaseVisionFaceDetectorOptions realtimeOpts =
                new FirebaseVisionFaceDetectorOptions.Builder()
                        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
                        .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
                        .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
                        .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
                        .build();

        detector = FirebaseVision.getInstance().getVisionFaceDetector(realtimeOpts);
        minSmileProb = 1;
        maxSmileProb = 0;
    }

    private void setupUI() {
        cameraView = findViewById(R.id.camera_front);
        tvAction = findViewById(R.id.tvAction);
        roundedFrameLayout = findViewById(R.id.main_frame);
    }*/
}

File: app/src/main/java/com/dev/fitface/ui/activity/CameraActivity.kt
inpackage com.dev.fitface.ui.activity

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.*
import android.media.Image
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.util.Log
import android.view.OrientationEventListener
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageProxy
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.dev.fitface.R
import com.dev.fitface.api.ApiService
import com.dev.fitface.camerax.CameraManager
import com.dev.fitface.interfaces.CameraCallback
import com.dev.fitface.mlkit.*
import com.dev.fitface.models.requests.FaceRequest
import com.dev.fitface.models.response.FaceResponse
import com.dev.fitface.ui.CustomToast
import com.dev.fitface.utils.*
import kotlinx.android.synthetic.main.activity_camera.*
import okhttp3.MediaType
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.create
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.ByteArrayOutputStream
import java.util.*
import kotlin.collections.ArrayList


class CameraActivity : AppCompatActivity() {

    private lateinit var cameraManager: CameraManager
    private lateinit var service: ApiService
    private var mCallback: CameraCallback? = null
    private var base64List: ArrayList<String>? = null
    private var roomId: String? = null
    private var token: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_camera)
        initValue()
        initService()
        createCameraManager()
        askPermission()
    }

    private fun initValue() {
        roomId = intent.getStringExtra("room_checkin")
        tvRoomID.text = roomId
        token = SharedPrefs.instance["Token", String::class.java]
        base64List = arrayListOf()
    }

    init {
        mCallback = object : CameraCallback {
            override fun onFaceCapture(rect: Rect) {
                tvAction.text = "Đã điểm danh"
                captureFace(rect)
                checkin(base64List!!, "CNTT3", roomId!!)

/*                cameraManager.stopCamera()
                Handler(Looper.getMainLooper()).postDelayed({
                    //Do something after 5000ms
                    cameraManager.startCamera()
                }, 2000)*/
                /*  imgCheckIn.setImageBitmap(a.base64ToImage(a))
                  imgCheckIn.visibility = View.VISIBLE

                  Handler(Looper.getMainLooper()).postDelayed({
                      //Do something after 5000ms
                      imgCheckIn.visibility = View.INVISIBLE
                      cameraManager.startCamera()
                      a = ""
                  }, 5000)*/
            }

            override fun onFaceSizeNotify(size: FaceSize) {
                when (size) {
                    FaceSize.SMALL -> {
                        tvAction.text = " Vui lòng tiến gần thiết bị hơn"
                    }
                    FaceSize.BIG -> {
                        tvAction.text = "Vui lòng cách xa thiết bị hơn"
                    }
                }
            }

            override fun onFrontFaceNotify() {
                tvAction.text = "Vui lòng nhìn thẳng vào thiết bị"
            }

            override fun onFaceOutsideNotify() {
                tvAction.text = "Vui lòng đặt khuôn mặt vào vùng hiển thị"
            }

            override fun onFaceInvalidNumber() {
                tvAction.text = "Có nhiều hơn một người"
            }

            override fun onFaceNone() {
                tvAction.text = ""
            }

            override fun onEyeNotify(status: EyeStatus) {
                when (status) {
                    EyeStatus.LEFT_EYE_CLOSE -> {
                        tvAction.text = "Vui lòng mở mắt trái"
                    }
                    EyeStatus.RIGHT_EYE_CLOSE -> {
                        tvAction.text = "Vui lòng mở mắt phải"
                    }
                    EyeStatus.ALL_EYES_CLOSE -> {
                        tvAction.text = "Vui lòng mở hai mắt"
                    }
                }
            }

        }
    }

    private fun askPermission() {
        if (allPermissionsGranted()) {
            cameraManager.startCamera()
        } else {
            ActivityCompat.requestPermissions(
                    this,
                    REQUIRED_PERMISSIONS,
                    REQUEST_CODE_PERMISSIONS
            )
        }
    }

    private fun initService() {
        service = ApiService.create()
    }

    private fun createCameraManager() {
        cameraManager = CameraManager(
                this,
                cameraView,
                this,
                graphicOverlay,
                mCallback
        )
    }

    private fun captureFace(faceRect: Rect) {
        setOrientationEvent()
        cameraManager.imageCapture.takePicture(
                cameraManager.cameraExecutor,
                object : ImageCapture.OnImageCapturedCallback() {
                    @SuppressLint("UnsafeExperimentalUsageError")
                    override fun onCaptureSuccess(image: ImageProxy) {
                        image.image?.let {
                            convertImageTpBitmap(it, faceRect)
                        }
                        super.onCaptureSuccess(image)
                    }
                })
    }

    private fun convertImageTpBitmap(image: Image, faceRect: Rect) {
        lateinit var resultStr: String
        image.imageToBitmap()
                ?.rotateFlipImage(
                        cameraManager.rotation,
                        cameraManager.isFrontMode()
                )
                ?.let { bitmap ->
                    graphicOverlay.processCanvas.drawBitmap(
                            bitmap,
                            0f,
                            bitmap.getBaseYByView(
                                    cameraView,
                                    cameraManager.isHorizontalMode()
                            ),
                            Paint().apply {
                                xfermode = PorterDuffXfermode(PorterDuff.Mode.DST_OVER)
                            })

//                  Create bitmap and encode to Base64 string
                    val left = faceRect.left
                    val top = faceRect.top
                    val right = faceRect.right
                    val bottom = faceRect.bottom
                    val width = right - left
                    val height = bottom - top
                    var res = Bitmap.createBitmap(bitmap, left, top, width, height)
                    var byteArrayOutputStream = ByteArrayOutputStream();
                    res.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream);
                    var byteArray = byteArrayOutputStream.toByteArray()
                    resultStr = Base64.encodeToString(byteArray, Base64.DEFAULT)
                    base64List?.clear()
                    base64List?.add(resultStr)

/*
                    runOnUiThread {
                        base64List?.add(resultStr)
                        Log.i("Debug","${base64List?.size}")
                        checkin(base64List!!, "CNTT3", roomId!!)
                    }*/
                }

    }

    private fun checkin(base64Str: ArrayList<String>, collection: String, roomId: String) {
        // Test
        Log.i("Debug", "${base64Str.size}  -call")
        val faceReq = FaceRequest(base64Str, collection)
        service.postCheckin(roomId, token!!, faceReq).enqueue(object : Callback<FaceResponse?> {
            override fun onResponse(call: Call<FaceResponse?>, response: Response<FaceResponse?>) {
                if (response.isSuccessful) {
                    val body = response.body()
                    if (body != null) {
                        if (body.status == 200 && body.data.isNotEmpty()) {
                            body.data.forEach { res ->
                                val mssv = res.username
                                val firstname = res.firstname
                                val lastname = res.lastname

                                val welcomStr = "Xin chào $mssv - $firstname $lastname"
                                tvAction.text = welcomStr
                            }
                        }
                    } else {
                        tvAction.text = "Điểm danh không hợp lệ"
                    }
                }
                else {
                    tvAction.text = "Điểm danh lỗi"
                }
            }

            override fun onFailure(call: Call<FaceResponse?>, t: Throwable) {
                CustomToast.makeText(this@CameraActivity, "Error", 400, CustomToast.ERROR).show()
            }

        })
    }


    private fun setOrientationEvent() {
        val orientationEventListener = object : OrientationEventListener(this as Context) {
            override fun onOrientationChanged(orientation: Int) {
                val rotation: Float = when (orientation) {
                    in 45..134 -> 270f
                    in 135..224 -> 180f
                    in 225..314 -> 90f
                    else -> 0f
                }
                cameraManager.rotation = rotation
            }
        }
        orientationEventListener.enable()
    }

    override fun onRequestPermissionsResult(
            requestCode: Int, permissions: Array<String>, grantResults:
            IntArray
    ) {
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                cameraManager.startCamera()
            } else {
                Toast.makeText(this, "Permissions not granted by the user.", Toast.LENGTH_SHORT)
                        .show()
                finish()
            }
        }
    }

    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(baseContext, it) == PackageManager.PERMISSION_GRANTED
    }


    companion object {
        private const val TAG = "CameraActivity"
        private const val REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS = arrayOf(
                Manifest.permission.CAMERA
        )
    }

}